srcts/src/components/webcomponents/layoutColumns.ts

```diff
@@ -0,0 +1,543 @@
+   * requested, this function needs to know the number of children in the
+   * layout, so it can determine the best-fit column widths.
+   *
+   * @returns The resolved column width specification.
+   */
+  private _resolveColWidthsSpec(): BreakpointColumnSpec {
+    const allResolved = Array.from(this.colWidths.values()).every(
+      (val) => !isNA(val)
+    );
+
+    if (allResolved) {
+      return newBreakpointColumnSpec(this.colWidths as BreakpointMapResolved);
+    }
+
+    const resolved = new Map() as BreakpointMapResolved;
+
+    const allAutoFit = Array.from(this.colWidths.values()).every((val) =>
+      isNA(val)
+    );
+
+    const units = allAutoFit ? null : 12;
+    const nChildren = this.children.length;
+
+    for (const [breakName, colWidth] of this.colWidths) {
+      if (colWidth === null) {
+        const preferWider = ["sm", "md"].includes(breakName);
+        const bestFit = bestFitColumnWidths(nChildren, preferWider, units);
+
+        if (allAutoFit) {
+          this.colUnits = bestFit.units;
+        }
+
+        resolved.set(breakName, bestFit.widths);
+      } else {
+        resolved.set(breakName, colWidth);
+      }
+    }
+
+    return newBreakpointColumnSpec(resolved);
+  }
+
+  /**
+   * Applies the column width specification to the children of the custom element.
+   */
+  private _applyColWidthsSpec(): void {
+    if (!this.colWidthsSpec) {
+      this.colWidthsSpec = this._resolveColWidthsSpec();
+    }
+
+    if (!this.colWidthsSpec) {
+      throw new Error("Column widths must be specified.");
+    }
+
+    const children = this.children;
+    writeGridClasses(this.colWidthsSpec, children, this.colUnits);
+  }
+}
+
+/**
+ * The best-fit algorithm chooses the number of column units per row and the
+ * widths of each column. These values are equivalent to user input.
+ */
+type BestFitColumnWidths = {
+  units: number;
+  widths: number[];
+};
+
+/**
+ * Given the number of children in the layout, choose the best-fit column widths
+ * for the layout. If `preferWider` is `true`, we prefer wider columns (fewer
+ * items per row, by default at the "sm" and "md" breakpoints), otherwise we
+ * prefer narrower columns (more items per row on larger screens).
+ *
+ * If all breakpoints are auto-fit, `units` will be `null`, allowing this
+ * function to choose the number of column units per row. This lets us get
+ * better results for small and irregular `nItems` that aren't possible with the
+ * 12-column grid. E.g. for 7 items we can use 14 units per row, with 7 column
+ * units per item on small screens and 2 column units per item on large screens.
+ * This choice is currently invariant to `nItems`.
+ *
+ * @param nItems The number of children in the layout.
+ * @param preferWider Whether to prefer wider columns (for smaller breakpoints).
+ * @param units The number of column units in a row.
+ * @returns The best-fit column widths.
+ */
+function bestFitColumnWidths(
```

Comment:
These helper functions could be methods on the `BslibLayoutColumns` class, but having started their life as R functions they're pure and don't depend on the state of a class instance. 

I felt like it was a little easier to see the lifecycle of a layout column class instance if these were not cluttering the class definition, but I'm ambivalent if you have strong opinions either way.